packages/augur-core/source/contracts/BaseSimpleDex.sol

```diff
@@ -0,0 +1,182 @@
+        return getSaleProceeds(_tokenAmount, tokenReserve, cashReserve);
+    }
+
+    function getCashSaleProceeds(uint256 _cashAmount) public view returns (uint256) {
+        return getSaleProceeds(_cashAmount, cashReserve, tokenReserve);
+    }
+
+    function getSaleProceeds(uint256 _inputAmount, uint256 _inputReserve, uint256 _outputReserve) public pure returns (uint256) {
+        require(_inputReserve > 0 && _outputReserve > 0, "inputReserve & outputReserve must be > 0");
+        uint256 _amountInputWithFee = _inputAmount.mul(997);
+        uint256 _numerator = _amountInputWithFee.mul(_outputReserve);
+        uint256 _denominator = _inputReserve.mul(1000).add(_amountInputWithFee);
+        return _numerator / _denominator;
+    }
+
+    function getTokenPurchaseCost(uint256 _tokenAmount) public view returns (uint256) {
+        return getPurchaseCost(_tokenAmount, cashReserve, tokenReserve);
+    }
+
+    function getCashPurchaseCost(uint256 _cashAmount) public view returns (uint256) {
+        return getPurchaseCost(_cashAmount, tokenReserve, cashReserve);
+    }
+
+    function getPurchaseCost(uint256 _outputAmount, uint256 _inputReserve, uint256 _outputReserve) public pure returns (uint256) {
+        require(_inputReserve > 0 && _outputReserve > 0, "inputReserve & outputReserve must be > 0");
+        uint256 _numerator = _outputAmount.mul(_inputReserve).mul(1000);
+        uint256 _denominator = _outputReserve.sub(_outputAmount).mul(997);
+        return _numerator / _denominator;
+    }
+
+    function autoSellToken(address _recipient, uint256 _tokenAmount) external payable returns (uint256 _cashAmount);
+
+    function sellToken(address _recipient) public nonReentrant returns (uint _cashAmount) {
+        uint256 _tokenBalance = getTokenBalance();
+        uint256 _tokenAmount = _tokenBalance.sub(tokenReserve);
+
+        _cashAmount = getTokenSaleProceeds(_tokenAmount);
+        require(_cashAmount > 0, "cashAmount must be > 0");
+        transferCash(_recipient, _cashAmount);
+
+        update(_tokenBalance, getCashBalance());
+        emit Swap(msg.sender, token, _tokenAmount, _cashAmount, _recipient);
+    }
+
+    function autoBuyToken(address _recipient, uint256 _cashAmount) external returns (uint256 _tokenAmount) {
+        augur.trustedCashTransfer(msg.sender, address(this), _cashAmount);
+        buyToken(_recipient);
+    }
+
+    function buyToken(address _recipient) public nonReentrant returns (uint256 _tokenAmount) {
+        uint256 _cashBalance = getCashBalance();
+        uint256 _cashAmount = _cashBalance.sub(cashReserve);
+
+        _tokenAmount = getCashSaleProceeds(_cashAmount);
+        require(_tokenAmount > 0, "tokenAmount must be > 0");
+        transferToken(_recipient, _tokenAmount);
+
+        update(getTokenBalance(), _cashBalance);
+        emit Swap(msg.sender, address(cash), _cashAmount, _tokenAmount, _recipient);
```

Comment:
During Maker global shutdown, the swap may be between vDAI and the token, rather than between DAI (cash) and the token.  In such a scenario, this event will produce incorrect results.

Reply:
because of what was noted above concerning not using `cashBalance` correct?

Reply:
In short, yes.  This function may transfer DAI or it may transfer vDAI, depending on the state of Maker.